src/gt4sd/algorithms/generation/moler/implementation.py

```diff
@@ -0,0 +1,99 @@
+#
+# MIT License
+#
+# Copyright (c) 2022 GT4SD team
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+"""Implementation of MoLeR conditional generators."""
+
+import logging
+from itertools import cycle, islice
+from typing import List
+
+from rdkit import Chem
+from molecule_generation import ModelWrapper
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+
+class MoLeRGenerator:
+    """Interface for MoLeR generator."""
+
+    def __init__(
+        self,
+        resources_path: str,
+        scaffolds: str,
+        num_samples: int,
+        beam_size: int,
+        seed: int,
+        num_workers: int,
+    ) -> None:
+        """Instantiate a MoLeR generator.
+
+        Args:
+            resources_path: path to the resources for model loading.
+            scaffolds: scaffolds as '.'-separated SMILES. If empty, no scaffolds are used.
+            num_samples: Number of molecules to sample per call.
+            beam_size: beam size to use during decoding.
+            seed: seed used for random number generation.
+            num_workers: number of workers used for generation.
+
+        Raises:
+            RuntimeError: in the case extras are disabled.
+        """
+        # loading artifacts
+        self.resources_path = resources_path
+        self.scaffolds = scaffolds
+        self.num_samples = num_samples
+        self.beam_size = beam_size
+        self.num_workers = num_workers
+        self._seed = seed
+
+    def generate(self) -> List[str]:
+        """Sample molecules using MoLeR.
+
+        Returns:
+            sampled molecule (SMILES).
+        """
+        # process scaffolds
+        valid_scaffolds = [
```

Comment:
This only makes sure the scaffolds are molecules but not that they are scaffolds.
We could use:

```py
if (
    Chem.MolToSmiles(Chem.MolFromSmiles(scaffold), canonical=True) ==
    Chem.Scaffolds.MurckoScaffold.MurckoScaffoldSmiles(scaffold)
)
```

I think the murcko scaffold canonicalizes under the hood so we need to account for that too,

Reply:
I think they have an imprecise naming in their function, they indeed accept any mol object as scaffolds following their call stack

Reply:
Looking at their paper, this is a bit surprising but we should probably stick to the original implementation.